attestation-agent/kbs_protocol/src/client/background_check.rs

```diff
@@ -0,0 +1,270 @@
+            version: String::from(KBS_PROTOCOL_VERSION),
+            tee: self._tee.clone().unwrap(),
+            extra_params: String::new(),
+        };
+
+        debug!("send auth request to {auth_endpoint}");
+
+        let challenge = self
+            .http_client
+            .post(auth_endpoint)
+            .header("Content-Type", "application/json")
+            .json(&request)
+            .send()
+            .await?
+            .json::<Challenge>()
+            .await?;
+
+        debug!("get challenge: {challenge:#?}");
+        let tee_pubkey = self.tee_key.export_pubkey()?;
+        let materials = vec![tee_pubkey.k_mod.as_bytes(), tee_pubkey.k_exp.as_bytes()];
+        let evidence = self.generate_evidence(challenge.nonce, materials).await?;
+        debug!("get evidence with challenge: {evidence}");
+
+        let attest_endpoint = format!("{}/{KBS_PREFIX}/attest", self.kbs_host_url);
+        let attest = Attestation {
+            tee_pubkey,
+            tee_evidence: evidence,
+        };
+
+        debug!("send attest request.");
+        let attest_response = self
+            .http_client
+            .post(attest_endpoint)
+            .header("Content-Type", "application/json")
+            .json(&attest)
+            .send()
+            .await?;
+
+        match attest_response.status() {
+            reqwest::StatusCode::OK => {
+                let resp = attest_response.json::<AttestationResponseData>().await?;
+                let token = Token::new(resp.token)?;
+                self.token = Some(token);
+            }
+            reqwest::StatusCode::UNAUTHORIZED => {
+                let error_info = attest_response.json::<ErrorInformation>().await?;
+                bail!("KBS attest unauthorized, Error Info: {:?}", error_info);
+            }
+            _ => {
+                bail!(
+                    "KBS Server Internal Failed, Response: {:?}",
+                    attest_response.text().await?
+                );
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn generate_evidence(&self, nonce: String, key_materials: Vec<&[u8]>) -> Result<String> {
+        let mut hasher = Sha384::new();
+        hasher.update(nonce.as_bytes());
+        key_materials
+            .iter()
+            .for_each(|key_material| hasher.update(key_material));
+
+        let ehd = hasher.finalize().to_vec();
+
+        let tee_evidence = self
+            .provider
+            .get_evidence(ehd)
+            .await
+            .context("Get TEE evidence failed")?;
+
+        Ok(tee_evidence)
+    }
+}
+
+#[async_trait]
+impl KbsClientCapabilities for KbsClient<Box<dyn EvidenceProvider>> {
+    async fn get_resource(&mut self, resource_uri: ResourceUri) -> Result<Vec<u8>> {
+        let remote_url = format!(
+            "{}/{KBS_PREFIX}/resource/{}/{}/{}",
+            self.kbs_host_url, resource_uri.repository, resource_uri.r#type, resource_uri.tag
+        );
+
+        if self._tee.is_none() {
```

Comment:
Maybe we should be checking `self.token` here instead? Notice that we make this exact same check as soon as we enter `rcar_handshake`.

Reply:
@Xynnn007 I agree, checking `self.token` here is better, and if `self.token.is_some()`, we needn't enter `rcar_handshake`.